packages/grid/veilid/server/veilid_streamer.py

```diff
@@ -0,0 +1,441 @@
+            3. Get STREAM_END request -> Reassemble message -> Send OK
+            ------ Operation for receiving the message finished here ------
+            4. Pass the reassembled message to the callback function and get the reply
+            5. Stream the reply back to the sender
+
+    Structs:
+        We are using 3 different structs to serialize and deserialize the metadata:
+
+        1. stream_start_struct = Struct("!8s16s32sQ")  # 64 bytes
+            [RequestType.STREAM_START (8 bytes string)] +
+            [Call ID (16 bytes random UUID string)] +
+            [Message hash (32 bytes string)] +
+            [Total chunks count (8 bytes unsigned long long)]
+
+        2. stream_chunk_header_struct = Struct("!8s16sQ")  # 32 bytes
+            [RequestType.STREAM_CHUNK (8 bytes string)] +
+            [Call ID (16 bytes random UUID string)] +
+            [Current Chunk Number (8 bytes unsigned long long)]
+
+        3. stream_end_struct = Struct("!8s16s")  # 24 bytes
+            [RequestType.STREAM_END (8 bytes string)] +
+            [Call ID (16 bytes random UUID string)]
+
+        The message is divided into chunks of 32736 bytes each, and each chunk is sent
+        as a separate STREAM_CHUNK request. This helps in keeping the size of each
+        request within the 32KB limit of the Veilid API.
+            [stream_chunk_header_struct (32 bytes)] +
+            [Actual Message Chunk (32736 bytes)]
+            = 32768 bytes
+
+    Usage:
+        1. Add this singleton class anwhere in your code, preferably above the update
+           callback function for your connection.
+           ```
+           vs = VeilidStreamer()
+           ```
+
+        2. Add a callback function to handle the received message stream:
+           ```
+           async def handle_receive_stream(message: bytes) -> bytes:
+               # Do something with the message once the entire stream is received.
+               return b'some response to the sender of the stream.'
+           ```
+
+        3. Add the following to your connection's update_callback function to relay
+           updates to the VeilidStreamer properly:
+           ```
+           def update_callback(update: veilid.VeilidUpdate) -> None:
+               if vs.is_stream_update(update):
+                   vs.receive_stream(connection, update, handle_receive_stream)
+               ...other callback code...
+           ```
+
+        4. Use the `stream` method to send an app_call with a message of any size.
+           ```
+           response = await vs.stream(router, dht_key, message)
+           ```
+    """
+
+    _instance = None
+    buffers: dict[CallId, Buffer]
+
+    def __new__(cls) -> "VeilidStreamer":
+        if cls._instance is None:
+            cls._instance = super().__new__(cls)
+            cls._instance.buffers = {}
+        return cls._instance
+
+    def __init__(self) -> None:
+        self.chunk_size = MAX_MESSAGE_SIZE
```

Comment:
MAYBE: we could rename this variable to 
```python3
self.max_message_size 
```
because in other places we chunk_size for formula calculation.

Reply:
Sure, that'll make more sense! How about renaming it to `self.max_single_message_size`?

Reply:
Because I think the former might imply a global maximum message size limit...